### PR TITLE
Fix canceling subscription in Watch

### PR DESCRIPTION
**Pull request by DanG100**

### CODE REVIEW

File: ygnmi/ygnmi.go [hunk 1]
```diff
@@ -313,25 +315,27 @@ func (w *Watcher[T]) Await() (*Value[T], error) {
 // Calling Await on the returned Watcher waits for the subscription to complete.
 // It returns the last observed value and a boolean that indicates whether that value satisfies the predicate.
 func Watch[T any](ctx context.Context, c *Client, q SingletonQuery[T], pred func(*Value[T]) error, opts ...Option) *Watcher[T] {
+	wCtx, cancel := context.WithCancel(ctx)
```
Comment (greg-dennis) [inline ygnmi/ygnmi.go:318]: more robust just to reassign ctx?

File: ygnmi/ygnmi.go [hunk 2]
```diff
@@ -313,25 +315,27 @@ func (w *Watcher[T]) Await() (*Value[T], error) {
...
 	resolvedOpts := resolveOpts(opts)
-	sub, err := subscribe[T](ctx, c, q, gpb.SubscriptionList_STREAM, resolvedOpts)
+	sub, err := subscribe[T](wCtx, c, q, gpb.SubscriptionList_STREAM, resolvedOpts)
 	if err != nil {
 		w.errCh <- err
 		return w
 	}
 
-	dataCh, errCh := receiveStream[T](ctx, sub, q)
+	dataCh, errCh := receiveStream[T](wCtx, sub, q)
```
Comment (greg-dennis) [inline ygnmi/ygnmi.go:331]: Can the cancel just be scoped to the goroutine, with no need to store anything in the watcher. Does this work?
ctx, cancel := context.WithCancel(ctx)
dataCh, errCh := receiveStream[T](wCtx, sub, q)
go func() {
  defer cancel()
  ...
}()

That would also mean the TCP session ends even if, for whatever reason, Await() is never called.

**Review by greg-dennis — CHANGES_REQUESTED** [review]

File: ygnmi/ygnmi.go (hunk 1 above)
Comment (DanG100) [reply ygnmi/ygnmi.go]: done

File: ygnmi/ygnmi.go (hunk 2 above)
Comment (DanG100) [reply ygnmi/ygnmi.go]: sure, just have to cancel in a few more place to make we don't leak the context

**Review by wenovus — APPROVED** [review]
So looks like we're saying we need the two goroutines to be able to cancel each other to avoid unnecessary work?

File: ygnmi/ygnmi.go [hunk 3]
```diff
@@ -313,19 +313,23 @@ func (w *Watcher[T]) Await() (*Value[T], error) {
...
+	var cancel context.CancelFunc
+	ctx, cancel = context.WithCancel(ctx)
 	w := &Watcher[T]{
 		errCh: make(chan error, 1),
 	}
 
 	resolvedOpts := resolveOpts(opts)
 	sub, err := subscribe[T](ctx, c, q, gpb.SubscriptionList_STREAM, resolvedOpts)
 	if err != nil {
+		cancel()
```
Comment (greg-dennis) [inline ygnmi/ygnmi.go:325]: This doesn't hurt anything, but not necessary, right? When a streaming RPC returns a non-nil error, it necessarily ends. The only case we technically need it is when we return early without having received an error. If that's right, you could delay the WithCancel call until right before the receiveStream call, and just leave the "defer cancel()" where it is.

**Review by greg-dennis — APPROVED** [review]

File: ygnmi/ygnmi.go (hunk 3 above)
Comment (DanG100) [reply ygnmi/ygnmi.go:325]: this was mostly to keep the linter happy.

**Comment by DanG100** [comment]
> So looks like we're saying we need the two goroutines to be able to cancel each other to avoid unnecessary work?

We need to cancel the context passed to the subscribe RPC so that stream actually closes

**Review by greg-dennis — APPROVED** [review]

**Pull request merged**